crates/holochain/src/core/workflow/incoming_dht_ops_workflow.rs

```diff
@@ -130,48 +108,51 @@ pub async fn incoming_dht_ops_workflow(
         dht_db,
         ..
     } = space;
-    let mut filter_ops = Vec::new();
-    let mut hashes_to_remove = Vec::with_capacity(ops.len());
 
-    // Filter out ops that are already being tracked, so we don't do duplicate work
-    {
-        let mut set = incoming_op_hashes.0.lock();
-        let mut o = Vec::with_capacity(ops.len());
-        for (hash, op) in ops {
-            if !set.contains(&hash) {
-                set.insert(hash.clone());
-                hashes_to_remove.push(hash.clone());
-                o.push((hash, op));
-            }
-        }
-        ops = o;
-    }
+    // Compute hashes for all the ops
+    let ops = ops
+        .into_iter()
+        .map(DhtOpHashed::from_content_sync)
+        .collect();
 
+    // Filter out ops that are already being tracked, to avoid doing duplicate work
+    let (_claim, ops) = OpsClaim::acquire(incoming_op_hashes, ops);
+
+    // If everything we've been sent is already being worked on then this workflow run can be skipped
     if ops.is_empty() {
         return Ok(());
     }
 
-    if !request_validation_receipt {
-        ops = filter_existing_ops(&dht_db, ops).await?;
-    }
-
-    for (hash, op) in ops {
-        // It's cheaper to check if the op exists before trying
-        // to check the signature or open a write transaction.
-        match should_keep(&op).await {
-            Ok(()) => filter_ops.push((hash, op)),
+    let num_ops = ops.len();
+    let mut filter_ops = Vec::with_capacity(num_ops);
+    for op in ops {
+        // It's cheaper to check if the signature is valid before proceeding to open a write transaction.
+        match should_keep(&op.content).await {
+            Ok(()) => filter_ops.push(op),
             Err(e) => {
                 tracing::warn!(
-                    msg = "Dropping op because it failed counterfeit checks",
-                    ?op
+                    ?op,
+                    "Dropping batch of {} ops because the current op failed counterfeit checks",
+                    num_ops,
                 );
+                // TODO we are returning here without blocking this author?
```

Comment:
Mean to raise this too, I added a TODO here. If somebody sends an incorrectly signed op should they get blocked? Or is there another mechanism like warrants to deal with this?

Reply:
Should get caught by the [`counterfeit_check`](https://github.com/holochain/holochain/blob/da6105f2886b6ec0b4f9b3a6b2c43fcf77aa12c5/crates/holochain/src/core/workflow/sys_validation_workflow.rs#L479)

Reply:
Oh, I see, we'll never get to the counterfeit check because that's in the next workflow! So you're right, I don't think it will result in a proper validation response.

Yeah, we need to have some better way of dealing with this case, and other cases involving invalid ops. We don't want to have lots of distinct pathways for validation -- we already have 2, which is 1 too many. But we do need a way to either run this through validation immediately, or send it to the sys validation workflow while using the absolute minimum system resources to deal with it.

Reply:
It would be handy to have some place to send stuff we know is invalid, like a paper shredder, whose sole job is to do the actual blocking and "integrate as invalid". Perhaps another distinct workflow, and we can break that part out from the actual validation workflows which determine validity. Then we can deal with all of these situations uniformly.

I'm not even certain where the blocking does happen when something is marked invalid, at first glance.